### PR TITLE
Add support for dynamic tests

### LINKED ISSUE #58: Add mechanism for registering dynamic tests
## Further Resources

- Related to #14.

## Subtasks

- [ ] #278 
- [ ] #275 
- [ ] #280

### PR DESCRIPTION
**Pull request by marcphilipp**
This is a rebased and squashed version of `issue58-dynamic-tests-with-new-resolvers` to resolve #58.

Added:
- Support discovering TestFactoryTestDescriptors by their unique ID
- Support `Iterable` in additon to `Collection`, `Iterator`, and `Stream` for `@TestFactory` methods

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

### CODE REVIEW

File: documentation/src/test/java/example/DynamicTestsDemo.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+		tests.add("Hallo");
+		return tests;
+	}
+
+	@TestFactory
+	List<DynamicTest> dynamicTestsFromList() {
+		List<DynamicTest> tests = new ArrayList<>();
+
+		tests.add(new DynamicTest("succeedingTest", () -> Assertions.assertTrue(true, "succeeding")));
+		tests.add(new DynamicTest("failingTest", () -> Assertions.assertTrue(false, "failing")));
```
Comment (marcphilipp) [inline documentation/src/test/java/example/DynamicTestsDemo.java]: - [ ] The `assert` methods should be statically imported.

File: junit-engine-api/src/main/java/org/junit/gen5/engine/support/hierarchical/SingleTestExecutor.java [hunk 2]
```diff
@@ -21,14 +21,15 @@
 /**
  * @since 5.0
  */
-class SingleTestExecutor {
-	interface Executable {
+public class SingleTestExecutor {
```
Comment (marcphilipp) [inline junit-engine-api/src/main/java/org/junit/gen5/engine/support/hierarchical/SingleTestExecutor.java]: - [ ] This class needs an `@API` annotation.

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/MethodTestDescriptor.java [hunk 3]
```diff
@@ -175,17 +175,17 @@ private void invokeBeforeTestMethodCallbacks(ExtensionRegistry registry, TestExt
 				.forEach(extension -> executeAndMaskThrowable(() -> extension.beforeTestMethod(context)));
 	}
 
-	private void invokeTestMethod(ExtensionRegistry ExtensionRegistry, TestExtensionContext testExtensionContext,
+	protected void invokeTestMethod(JUnit5EngineExecutionContext context, TestExtensionContext testExtensionContext,
```
Comment (marcphilipp) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/MethodTestDescriptor.java]: - [ ] Why was this changed?

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/TestFactoryTestDescriptor.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	}
+
+	@Override
+	public boolean isContainer() {
+		return true;
+	}
+
+	@Override
+	public boolean isTest() {
+		return true;
```
Comment (marcphilipp) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/TestFactoryTestDescriptor.java]: - [ ] This should return `false`. However, then launcher will prune the test descriptor.

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/MethodTestDescriptor.java (hunk 3 above)
Comment (sbrannen) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/MethodTestDescriptor.java]: Actually, it looks like this method should not have been modified at all. The `context` is not used except to access the `ExtensionRegistry`. Thus, the changes look superfluous, perhaps simply due to multiple merges with master over time.

Comment (marcphilipp) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/MethodTestDescriptor.java]: Yeah, I think so, too. I'll remove this change.

Comment (marcphilipp) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/MethodTestDescriptor.java]: Ah, never mind: It was changed so `TestFactoryTestDescriptor` can override the method.

File: documentation/src/test/java/example/DynamicTestsDemo.java (hunk 1 above)
Comment (marcphilipp) [inline documentation/src/test/java/example/DynamicTestsDemo.java]: I've added a subtask to #275 to address this.

File: junit-engine-api/src/main/java/org/junit/gen5/engine/support/hierarchical/SingleTestExecutor.java (hunk 2 above)
Comment (marcphilipp) [inline junit-engine-api/src/main/java/org/junit/gen5/engine/support/hierarchical/SingleTestExecutor.java]: Done in d51f874a1250f2ebde26f0963f29a6f4fdab75fb.

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/TestFactoryTestDescriptor.java (hunk 4 above)
Comment (marcphilipp) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/TestFactoryTestDescriptor.java]: I've added an issue to address this: #280.

**Comment by marcphilipp** [comment]
I think this pull request is good to be merged now. The remaining work will be addressed in #275 and #280.

@junit-team/junit-lambda Please take a look and approve. 😄

**Comment by sbrannen** [comment]
@marcphilipp, I think this looks good for an initial inclusion in `master`.

So... go for it!

**Comment by marcphilipp** [comment]
This pull request has been merged into master.

**Comment by sbrannen** [comment]
👍 

@jlink & @mmerdes, thanks for implementing this much needed feature!

@marcphilipp , thanks for getting a clean merge on top of `master`!
